### ISSUE TITLE
MIxing versions leads to subtle breakage, please version your protocol

### ISSUE THREAD

**Issue by deliciouslytyped** [open-iscsi/open-iscsi#222]
This is not a typical problem on most popular systems, but NixOS (and an increasing amount of software with mechanisms resembling it) allows one to effortlessly - and so accidentally - install or use multiple versions of software. This can lead to subtle breakage with open-iscsi tools.

I humbly propose that the project should do something about versioning their communication protocol to prevent issues like https://github.com/NixOS/nixpkgs/issues/97875 .

A somewhat similar example where they do some manner of checking between the client and "server" components is the zfs kernal and userspace tools.

**Comment by deliciouslytyped** [comment]
Here are straces of the daemon and client I still happen to have handy:
[daemon.trace.txt](https://github.com/open-iscsi/open-iscsi/files/5214144/daemon.trace.txt)
[client.trace.txt](https://github.com/open-iscsi/open-iscsi/files/5214145/client.trace.txt)

**Comment by gonzoleeman** [comment]
Marking this as a request for enhancement. Adding such a mechanism would of course also be an incompatible change, i.e. code from before that change would not work with code after that change.

**Comment by deliciouslytyped** [comment]
I don't know how the open-iscsi protocol or unix sockets work, but if it can handle version queries from/to the client/server in packets separate from the normal operations, it should be possible to implement in a backwards compatible manner if the current implementation doesn't crash on unknown commands?

This approach would again allow for version mismatches if the check is only done once at startup time (and say someone (re)starts a different server, or client version), so the check should be run before every operation? This should work if packet order is preserved?

**Comment by gonzoleeman** [comment]
Yes, checking before each communication to see if the version matches would be wasteful and therefore impractical.

Ideally, the data structures passed back and forth would be versioned, or else the protocol would in general have a field for "version" in each communication.

Also, the communication now between the daemon and the kernel is more like RPC, where we pass a function "number", and the other side uses that to index into a list of functions. So both the kernel and the daemon need that to match, I believe.

**Comment by deliciouslytyped** [comment]
Pardon my ignorance, I'm not really much of a systems programmer;
It's not _very_ wasteful if its "just" going over a local unix socket? besides, how bad can it be? It shouldn't be more than a couple bytes back and forth?
If you wanted to be a bit more conservative you could do something like sending a message every second before/during activity or somesuch? But that doesn't seem worth the complexity compared to just checking on every message?

With respect to matching function calls back and forth, does the server just segfault if you call the wrong number? The client should detect that there's a version mismatch or otherwise something suspicious is happening, if it doesn't receive a reply to the version request function call?

**Comment by gonzoleeman** [comment]
Please feel free to submit a patch to version the communication, for both open-iscsi and the kernel, and I'll be glad to help you benchmark the effect, if any. It's true that normal IO does not go through the user-land/kernel interface, but setup and query does, and I have seen systems with thousands of targets over multiple paths. When a path fails and many targets fail over, there can be quite a bit of communication. I consider this an enhancement request, since the design of open-iscsi currently requires the kernel and user pieces to match. Or perhaps a design flaw? But I hate crappy fixes, so even if I wanted to handle version differences, I think I'd prefer to bite the one-time bullet of adding versioning to the data structures passed (ala RPC) rather than double our communication load. But testing might persuade me that either approach works.

P.S. If you're not a systems programmer what makes you think it's easy?

**Comment by deliciouslytyped** [comment]
That's why everything ends in a questionmark.

**Comment by gonzoleeman** [comment]
I spent some time looking at the nettlink code in open-iscsi yesterday in reference to this issue. "Non-trivial" certainly under describes the complexity.

There is an "iscsi_uevent" (iscsi user-land event?) structure that is passed to the kernel then returned, for things like session creation, etc, where iscsiadm asks the kernel code to do something.

But I'm not sure this will help you enough for your goal. This would just "fix" the iscsid <-> kernel communication. You still have a **lot** of other places where matches version are expected (i.e. interfaces). In the issue you referenced initially, the problem seems to be between iscsiadm and iscsid. So you'd have to version that, too, plus handle log for handling missing versions. I've worked on code before that handled older versions, and it's kind of a PITA, in that you sometimes have to add code like: "if current_version do_regular_stuff() else handle_older_versions()".

Oh, and there's our "database", i.e. a bunch of files under (usually) /etc/iscsi/nodes. Changes were made to this file for our recent CHAP enhancements, and older version of open-iscsi don't like the new format (which I think is what happened in your issue). In this case, it's impossible to go back and change the older open-iscsi code to understand the new format. Again, perhaps these "database" files should be versioned, which would fix this for _future_ versions, but in and of itself would be an incompatible change. :-/ 

In the case of open-iscsi, we have not tried to block incompatible changes because doing so stunts progress. We certainly never planned for incompatible versions of iscsiadm and iscsid!

One other note: there actually is a version check when iscsi talks to the kernel, but the versions haven't been upgraded in a while, since that interfaces has been fairly stable. Perhaps we could add a check when iscsiadm talks to iscsid that just bailed if the versions did not match, which is arguably better than hanging. I'm not sure if this would have caught your particular case or not.

**Comment by deliciouslytyped** [comment]
Unless I missed your point, I'm not saying old and new versions should be cross-compatible in general, I'm saying it _just_ in the case of the version check, which would be run before any other requests, and abort if there's a mismatch - or do you mean the database files are necessary to even get that far?

**Comment by gonzoleeman** [comment]
as I believe I mentioned before, there are several interfaces:

1. iscsid <-> kernel
2. iscsiadm <-> iscsid
3. iscsiadm <-> node database
4. iscsiadm <-> send_targets database
5. iscsiadm <-> sysfs

In your case, it's not clear to me which of these was the problems, as the details are a bit vague. Protecting each of these interfaces with version checks would be done differently.

If you can supply details so that I can reproduce the problem, I could at least see where the actual problem occurred. But I suspect that even if we fixed that issue, there are others waiting. Each of the listed interfaces really would need to be protected IMHO.

**Comment by deliciouslytyped** [comment]
In theory, I should be able to repro.
